File: java/src/main/java/com/logicalclocks/hsfs/engine/DeltaEngine.java
/*
 * Copyright (c) 2020 Logical Clocks AB
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *
 * See the License for the specific language governing permissions and limitations under the License.
 */

package com.logicalclocks.hsfs.engine;

import com.logicalclocks.hsfs.FeatureGroup;
import com.logicalclocks.hsfs.commit.DeltaLakeUpdate;
import com.logicalclocks.hsfs.commit.FeatureGroupCommit;
import com.logicalclocks.hsfs.FeatureStoreException;
import com.logicalclocks.hsfs.ActionType;
import com.logicalclocks.hsfs.metadata.FeatureGroupApi;

import io.delta.tables.DeltaMergeBuilder;
import io.delta.tables.DeltaMergeMatchedActionBuilder;
import io.delta.tables.DeltaMergeNotMatchedActionBuilder;
import io.delta.tables.DeltaTable;

import lombok.SneakyThrows;
import org.apache.spark.sql.Dataset;
import org.apache.spark.sql.Row;
import org.apache.spark.sql.SaveMode;
import org.apache.spark.sql.SparkSession;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import scala.collection.JavaConverters;

import java.io.IOException;
import java.sql.Timestamp;
import java.text.SimpleDateFormat;
import java.util.Date;
import java.util.HashMap;
import java.util.Map;

public class DeltaEngine {

  private static final String DELTA_SPARK_FORMAT = "delta";
  private SimpleDateFormat dateFormat = new SimpleDateFormat("yyyyMMddHHmmss.SSS");

  private static final Logger LOGGER = LoggerFactory.getLogger(DeltaEngine.class);

  private Utils utils = new Utils();
  private FeatureGroupApi featureGroupApi = new FeatureGroupApi();
  private FeatureGroupCommit fgCommitMetadata = new FeatureGroupCommit();

  public void saveDeltaLakeFeatureGroup(SparkSession sparkSession, FeatureGroup featureGroup, Dataset<Row> dataset,
                                        SaveMode saveMode, ActionType operation,
                                        Map<String, String> writeOptions, Map<String, Object> deltaCustomExpressions)
      throws FeatureStoreException, IOException {

    if (deltaCustomExpressions == null) {
      deltaCustomExpressions = new HashMap<>();
    }

    if (operation == ActionType.BULK_INSERT) {
      dataset.write()
          .format(DELTA_SPARK_FORMAT)
          .partitionBy(utils.getPartitionColumns(featureGroup))
          // write options cannot be null
          .options(writeOptions == null ? new HashMap<>() : writeOptions)
          .mode(saveMode)
          .save("hdfs:///apps/hive/warehouse/feltafs_featurestore.db/" + utils.getTableName(featureGroup));

      //dataset.write().format(DELTA_SPARK_FORMAT).saveAsTable(utils.getTableName(featureGroup));
      FeatureGroup apiFG = featureGroupApi.save(featureGroup);
      if (featureGroup.getVersion() == null) {
        LOGGER.info("VersionWarning: No version provided for creating feature group `" + featureGroup.getName()
            + "`, incremented version to `" + apiFG.getVersion() + "`.");
      }

      // Update the original object - Hopsworks returns the incremented version
      featureGroup.setId(apiFG.getId());
      featureGroup.setVersion(apiFG.getVersion());
      featureGroup.setLocation(apiFG.getLocation());
      featureGroup.setId(apiFG.getId());
      featureGroup.setCorrelations(apiFG.getCorrelations());
      featureGroup.setHistograms(apiFG.getHistograms());

    } else if (operation == ActionType.INSERT) {
      dataset.write()
          .format(DELTA_SPARK_FORMAT)
          // write options cannot be null
          .options(writeOptions == null ? new HashMap<>() : writeOptions)
          .mode(SaveMode.Append)
          .save(featureGroup.getLocation());
    } else if (operation == ActionType.UPSERT) {
      DeltaLakeUpdate deltaLakeMetaData = new DeltaLakeUpdate();
      String newDataAlias = utils.getFgName(featureGroup) + "_update";
      deltaLakeMetaData.setExistingAlias(utils.getFgName(featureGroup));
      deltaLakeMetaData.setNewDataAlias(newDataAlias);

      DeltaLakeUpdate updatedDeltaLakeMetaData = featureGroupApi.constructDeltaLakeUpdate(featureGroup,
            deltaLakeMetaData);

      DeltaMergeBuilder deltaMergeBuilder = DeltaTable.forPath(sparkSession, featureGroup.getLocation())
            .as(utils.getFgName(featureGroup))
            .merge(
                dataset.as(newDataAlias),
                deltaCustomExpressions.containsKey("mergeExp")
                    ? deltaCustomExpressions.get("mergeExp").toString()
                    : updatedDeltaLakeMetaData.getMergeCondQuery());

      DeltaMergeMatchedActionBuilder deltaMergeMatchedActionBuilder;
      if (writeOptions.containsKey("whenMatchedExp")) {
        deltaMergeMatchedActionBuilder = deltaMergeBuilder.whenMatched(
            deltaCustomExpressions.get("whenMatchedExp").toString());
      } else {
        deltaMergeMatchedActionBuilder = deltaMergeBuilder.whenMatched();
      }

      if (deltaCustomExpressions.containsKey("updateExp")) {
        deltaMergeBuilder = deltaMergeMatchedActionBuilder.updateExpr(
            (Map<String, String>) deltaCustomExpressions.get("updateExp"));
      } else {
        deltaMergeBuilder = deltaMergeMatchedActionBuilder.updateAll();
      }

      DeltaMergeNotMatchedActionBuilder deltaMergeNotMatchedActionBuilder;
      if (writeOptions.containsKey("whenMatchedExp")) {
        deltaMergeNotMatchedActionBuilder = deltaMergeBuilder.whenNotMatched(
            deltaCustomExpressions.get("whenMatchedExp").toString());
      } else {
        deltaMergeNotMatchedActionBuilder = deltaMergeBuilder.whenNotMatched();
      }


      if (deltaCustomExpressions.containsKey("insertExp")) {
        deltaMergeBuilder = deltaMergeNotMatchedActionBuilder.insertExpr(
            (Map<String, String>) deltaCustomExpressions.get("insertExp"));
      } else {
        deltaMergeBuilder = deltaMergeNotMatchedActionBuilder.insertAll();
      }

      deltaMergeBuilder.execute();
    }

    FeatureGroupCommit fgCommit = getLastCommitMetadata(sparkSession, featureGroup);
    featureGroupApi.featureGroupCommit(featureGroup, fgCommit);
  }

  // TODO (davit):
  public FeatureGroupCommit deleteRecord() {
    return null;
  }

  public void registerTemporaryTable(SparkSession sparkSession, FeatureGroup featureGroup, String alias,
                                     Long timestampAsOf, Map<String, String> userReadOptions) {
    Map<String, String> readOptions = new HashMap<>();
    readOptions.put("timestampAsOf", new Timestamp(timestampAsOf).toString());
    if (userReadOptions != null) {
      readOptions.putAll(userReadOptions);
    }
    sparkSession.read()
        .format(DELTA_SPARK_FORMAT)
        .options(readOptions)
        .load(featureGroup.getLocation())
        .createOrReplaceTempView(alias);
  }

  public FeatureGroupCommit getLastCommitMetadata(SparkSession sparkSession, FeatureGroup featureGroup) {
    Row lastCommit = DeltaTable.forPath(sparkSession, featureGroup.getLocation()).history(1).first();
    Long version = lastCommit.getAs("version");
    String timestamp = lastCommit.getAs("timestamp").toString();
    fgCommitMetadata.setCommitDateString(timestamp);
    fgCommitMetadata.setDeltaCommitVersion(version);

    Map<String, String> operationMetrics = JavaConverters.mapAsJavaMapConverter(
        (scala.collection.immutable.Map<String, String>) lastCommit.getAs("operationMetrics")).asJava();
    if (version == 0) {
      fgCommitMetadata.setRowsUpdated(0L);
      fgCommitMetadata.setRowsInserted(Long.parseLong(operationMetrics.get("numOutputRows")));
      fgCommitMetadata.setRowsDeleted(0L);
    } else {
      fgCommitMetadata.setRowsUpdated(Long.parseLong(operationMetrics.get("numTargetRowsUpdated")));
      fgCommitMetadata.setRowsInserted(Long.parseLong(operationMetrics.get("numTargetRowsInserted")));
      fgCommitMetadata.setRowsDeleted(Long.parseLong(operationMetrics.get("numTargetRowsDeleted")));
    }

    return fgCommitMetadata;
  }

  @SneakyThrows
  public String timeStampToCommitFormat(Long commitedOnTimeStamp) {
    Date commitedOnDate = new Timestamp(commitedOnTimeStamp);
    return dateFormat.format(commitedOnDate);
  }
}

File: python/hsfs/core/delta_engine.py
#
#   Copyright 2024 Hopsworks AB
#
#   Licensed under the Apache License, Version 2.0 (the "License");
#   you may not use this file except in compliance with the License.
#   You may obtain a copy of the License at
#
#       http://www.apache.org/licenses/LICENSE-2.0
#
#   Unless required by applicable law or agreed to in writing, software
#   distributed under the License is distributed on an "AS IS" BASIS,
#   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#   See the License for the specific language governing permissions and
#   limitations under the License.
#

from hsfs import feature_group_commit, util
from hsfs.core import feature_group_api

try:
    from delta.tables import DeltaTable
except ImportError:
    pass


class DeltaEngine:
    DELTA_SPARK_FORMAT = "delta"
    DELTA_QUERY_TIME_TRAVEL_AS_OF_INSTANT = "timestampAsOf"

    def __init__(
        self,
        feature_store_id,
        feature_store_name,
        feature_group,
        spark_session,
        spark_context,
    ):
        self._feature_group = feature_group
        self._spark_context = spark_context
        self._spark_session = spark_session
        self._feature_store_id = feature_store_id
        self._feature_store_name = feature_store_name

        self._feature_group_api = feature_group_api.FeatureGroupApi()

    def save_delta_fg(self, dataset, write_options, validation_id=None):
        fg_commit = self._write_delta_dataset(dataset, write_options)
        fg_commit.validation_id = validation_id
        return self._feature_group_api.commit(self._feature_group, fg_commit)

    def register_temporary_table(self, delta_fg_alias, read_options):
        delta_options = self._setup_delta_read_opts(delta_fg_alias, read_options)
        self._spark_session.read.format(self.DELTA_SPARK_FORMAT).options(
            **delta_options
        ).load(self._feature_group.location).createOrReplaceTempView(
            delta_fg_alias.alias
        )

    def _setup_delta_read_opts(self, delta_fg_alias, read_options):
        delta_options = {}
        if delta_fg_alias.left_feature_group_end_timestamp is None and (
            delta_fg_alias.left_feature_group_start_timestamp is None
            or delta_fg_alias.left_feature_group_start_timestamp == 0
        ):
            # snapshot query latest state
            delta_options = {}
        elif (
            delta_fg_alias.left_feature_group_end_timestamp is not None
            and delta_fg_alias.left_feature_group_start_timestamp is None
        ):
            # snapshot query with end time
            _delta_commit_end_time = util.get_delta_datestr_from_timestamp(
                delta_fg_alias.left_feature_group_end_timestamp
            )
            delta_options = {
                self.DELTA_QUERY_TIME_TRAVEL_AS_OF_INSTANT: _delta_commit_end_time,
            }

        if read_options:
            delta_options.update(read_options)

        return delta_options

    def delete_record(self, delete_df, write_options):
        pass

    def _write_delta_dataset(self, dataset, write_options):
        if write_options is None:
            write_options = {}

        if not DeltaTable.isDeltaTable(
            self._spark_session, self._feature_group.location
        ):
            (
                dataset.write.format(DeltaEngine.DELTA_SPARK_FORMAT)
                .options(**write_options)
                .partitionBy(
                    self._feature_group.partition_key
                    if self._feature_group.partition_key
                    else []
                )
                .mode("append")
                .save(self._feature_group.location)
            )
        else:
            fg_source_table = DeltaTable.forPath(
                self._spark_session, self._feature_group.location
            )

            source_alias = (
                f"{self._feature_group.name}_{self._feature_group.version}_source"
            )
            updates_alias = (
                f"{self._feature_group.name}_{self._feature_group.version}_updates"
            )
            merge_query_str = self._generate_merge_query(source_alias, updates_alias)

            fg_source_table.alias(source_alias).merge(
                dataset.alias(updates_alias), merge_query_str
            ).whenMatchedUpdateAll().whenNotMatchedInsertAll().execute()

        return self._get_last_commit_metadata(
            self._spark_session, self._feature_group.location
        )

    def _generate_merge_query(self, source_alias, updates_alias):
        merge_query_list = []
        primary_key = self._feature_group.primary_key

        # add event time to primary key for upserts
        if self._feature_group.event_time is not None:
            primary_key.append(self._feature_group.event_time)

        # add partition key for upserts
        if self._feature_group.partition_key:
            primary_key = primary_key + self._feature_group.partition_key

        for pk in primary_key:
            merge_query_list.append(f"{source_alias}.{pk} == {updates_alias}.{pk}")
        megrge_query_str = " AND ".join(merge_query_list)
        return megrge_query_str

    @staticmethod
    def _get_last_commit_metadata(spark_context, base_path):
        fg_source_table = DeltaTable.forPath(spark_context, base_path)
        last_commit = fg_source_table.history(1).first().asDict()
        version = last_commit["version"]
        commit_timestamp = util.convert_event_time_to_timestamp(
            last_commit["timestamp"]
        )
        commit_date_string = util.get_hudi_datestr_from_timestamp(commit_timestamp)
        operation_metrics = last_commit["operationMetrics"]

        if version == 0:
            fg_commit = feature_group_commit.FeatureGroupCommit(
                commitid=None,
                commit_date_string=commit_date_string,
                commit_time=commit_timestamp,
                rows_inserted=operation_metrics["numOutputRows"],
                rows_updated=0,
                rows_deleted=0,
                last_active_commit_time=commit_timestamp,
            )
        else:
            fg_commit = feature_group_commit.FeatureGroupCommit(
                commitid=None,
                commit_date_string=commit_date_string,
                commit_time=commit_timestamp,
                rows_inserted=operation_metrics["numTargetRowsInserted"],
                rows_updated=operation_metrics["numTargetRowsUpdated"],
                rows_deleted=operation_metrics["numTargetRowsDeleted"],
                last_active_commit_time=commit_timestamp,
            )

        return fg_commit
